plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -249,6 +249,189 @@ fu_uefi_setup_bootnext_with_loadopt(FuEfiLoadOption *loadopt, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_uefi_bootmgr_shim_is_safe(const gchar *source_shim, GError **error)
+{
+	g_autofree gchar *sbatlevel_path = NULL;
+	g_autofree gchar *sbatlevel_content_orig = NULL;
+	g_autofree gchar *sbatlevel_content = NULL;
+	gsize sbatlevel_size = 0;
+	g_autofree gchar *current_sbatlevel = NULL;
+	g_autofree gchar *current_sbatlevel_orig = NULL;
+	gsize current_sbatlevel_size = 0;
+
+	sbatlevel_path = g_strdup_printf("%s.sbatlevel", source_shim);
+	if (!g_file_test(sbatlevel_path, G_FILE_TEST_EXISTS)) {
+		/*
+		 * This file is expected in the snap.  Other
+		 * distributions of fwupd are probably safe since they
+		 * will use the host's shim to do an update.
+		 */
+		return TRUE;
+	}
+
+	if (!g_file_get_contents(sbatlevel_path, &sbatlevel_content_orig, &sbatlevel_size, error)) {
+		return FALSE;
+	}
+
+	/* Make it null terminated to make g_ascii_strtoll safe to call */
+	sbatlevel_content = g_realloc(sbatlevel_content_orig, sbatlevel_size + 1);
+	if (sbatlevel_content != NULL)
+		g_steal_pointer(&sbatlevel_content_orig);
+	g_assert(sbatlevel_content != NULL);
+	sbatlevel_content[sbatlevel_size] = '\0';
+
+	/* Not safe if variable is not set but new shim would set it */
+	if (!fu_efivar_get_data(FU_EFIVAR_GUID_SHIM,
+				"SbatLevelRT",
+				(guint8 **)&current_sbatlevel_orig,
+				&current_sbatlevel_size,
+				NULL,
+				NULL))
+		return FALSE;
+
+	/* Make it null terminated to make g_ascii_strtoll safe to call */
+	current_sbatlevel = g_realloc(current_sbatlevel_orig, current_sbatlevel_size + 1);
+	if (current_sbatlevel != NULL)
+		g_steal_pointer(&current_sbatlevel_orig);
+	g_assert(current_sbatlevel != NULL);
+	current_sbatlevel[current_sbatlevel_size] = '\0';
+
+	/*
+	 * For every new shim entry, we need a matching entry in the
+	 * current sbatlevel. That is the entry of the shim is not
+	 * newer than current sbatlevel.
+	 *
+	 * The opposite way might work (for example shim's latest
+	 * sbatlevel matches) or not (shim is too old), but it will
+	 * not brick the current OS.
+	 */
+	for (size_t shim_entry = 0; sbatlevel_content[shim_entry];) {
```

Comment:
can we maybe make this a lot safer and split by line, and then split each line by comma? There's a lot of scary pointer accesses going on.

Reply:
OK, will do that.

Reply:
This is done.